R/unnest.R

```diff
@@ -0,0 +1,245 @@
+# Expand a template string containing `{col}` for each column name
+#
+# @param template A string that may contain `{col}` as a placeholder.
+# @param col_names A character vector of column names to substitute.
+#
+# @return If template is NULL, returns NULL. If template contains `{col}`,
+#   returns a named character vector where values are the expanded names and
+#   names are the original column names. Otherwise returns the template as-is.
+#
+# @noRd
+expand_col_template <- function(template, col_names) {
+  if (is.null(template)) {
+    return(NULL)
+  }
+
+  is_template <- grepl("\\{col\\}", template)
+
+  if (is_template) {
+    result <- vapply(
+      col_names,
+      function(x) gsub("\\{col\\}", x, template),
+      FUN.VALUE = character(1)
+    )
+    names(result) <- col_names
+    return(result)
+  }
+
+  template
+}
+
+# Check if a string is a `{col}` template
+#
+# @param x A string or NULL.
+#
+# @return TRUE if x contains `{col}`, FALSE otherwise.
+#
+# @noRd
+is_col_template <- function(x) {
+  !is.null(x) && grepl("\\{col\\}", x)
+}
+
+
+#' Unnest a list-column into rows
```

Comment:
Nitpick: I prefer the exported function to be at the top of the file and the helper functions below it.